### ISSUE TITLE
Type narrowing on some logic isn't as good as it could be

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#4643]
Consider this code:
```python
from typing import Optional

def foo(x1: Optional[int], x2: Optional[int]) -> int:

    if x1 is None and x2 is None:
        return 0
    if x1 is not None and x2 is not None:
        return x1 + x2
    else:
        s = x1 or x2
        reveal_type(s)
        return s
```

pyright 1.1.293 reports this:
```text
  inferpairs.py:11:21 - information: Type of "s" is "int | None"
  inferpairs.py:12:16 - error: Expression of type "int | None" cannot be assigned to return type "int"
    Type "int | None" cannot be assigned to type "int"
      Type "None" cannot be assigned to type "int" (reportGeneralTypeIssues)
```

But if you apply some boolean logic, you know that `s` cannot be `None`.

Not sure if this is something `pyright` can detect or not.

**Comment by erictraut** [comment]
Type narrowing is applied on individual symbols, not conditional pairs of symbols. To do what you're suggesting above, pyright would need to track that "the type of `x1` is `None` if `x2` is also `None` but otherwise some other type. That sort of conditional type tracking would be too complex and computationally infeasible in a static type checker. I don't know if any type checker or compiler in any language that does this. We certainly don't plan to support this.

There's almost always a way to refactor your code to accommodate type narrowing.

```python
def foo(x1: Optional[int], x2: Optional[int]) -> int:
    if x1 is not None and x2 is not None:
        return x1 + x2
    elif x1 is None:
        return x2 or 0
    else:
        return x1
```
